File: Sort/插入排序/InsertSort.java
/*
直接插入排序
*/
void insertSort(int[] arr) {
	for (int i=1; i<arr.length; i++) {
		int key = arr[i], j;
		for (j=i-1; j>=0 && arr[j]>key; j--)
			arr[j+1] = arr[j];
		arr[j+1] = key;
	}
}

/*
折半插入排序
*/
void binaryInsertSort(int arr[]) {
	for (i=1; i<arr.length; i++) {
		int L=0, R=i-1;
		int key = arr[i];
		while (L <= R) {
			int m = L + (R-L)/2;
			if (arr[m] > arr[i]) {
				R = m - 1;
			} else {
				L = m + 1;
			}
		}
		for (int j=i-1; j >= L; j--) {
			arr[j+1] = arr[j];
		}
		arr[L] = key;
	}
}

/*
希尔排序
*/

File: Sort/插入排序/InsertSort.cpp
//A[0]为哨兵

/*
直接插入排序
*/
void insertSort(ElemType A[], int n)
{
	int i, j;
	for (i=2; i<=n; i++)
	{
		if (A[i].key < A[i-1].key)
		{
			A[0] = A[i];
			for (j=i-1; A[0].key<A[j].key; --j)//比A[j]大的往后移
			{
				A[j+1] = A[j];
			}
			A[j+1] = A[0];
		}
	}
}

/*
折半插入排序
*/
void binaryInsertSort(ElemType A[], int n)
{
	int i, j, low, high, mid;
	for (i=2; i<=n; i++) {
		A[0] = A[i];
		low = 1;  high = i-1;
		while (low <= high) {
			mid = (low + high) / 2;
			if (A[mid].key > A[0].key) 
				high = mid - 1;
			else 
				low = mid + 1;
		}
		for (j=i-1; j>=high+1; --j) {
			A[j+1] = A[j];
		}
		A[high+1] = A[0];
	}
}

/*
希尔排序
*/